File: go/internal/booking-service/core/domain/entities/booking.entity.go
package entities

import (
	"book/internal/booking-service/core/domain/enums"
	"time"
)

var transitions = map[struct {
	status enums.Status
	action enums.Actions
}]enums.Status{
	{enums.Created, enums.Pay}:     enums.Paid,
	{enums.Created, enums.Cancel}:  enums.Canceled,
	{enums.Paid, enums.Finish}:     enums.Finished,
	{enums.Paid, enums.Refound}:    enums.Refounded,
	{enums.Canceled, enums.Reopen}: enums.Created,
}

type Booking struct {
	Id       uint `gorm:"primaryKey"`
	PlacedAt time.Time
	Start    time.Time
	End      time.Time
	status   enums.Status
	GuestID  uint `gorm:"not null"`
	RoomID   uint `gorm:"not null"`
}

func (b *Booking) CurrentStatus() enums.Status {
	return b.status
}

func (b *Booking) ChangeState(action enums.Actions) {
	transition := struct {
		status enums.Status
		action enums.Actions
	}{b.status, action}

	if newStatus, exists := transitions[transition]; exists {
		b.status = newStatus
	}
}

func NewBooking() *Booking {
	return &Booking{
		status: enums.Created,
	}
}

File: nestjs/src/booking/core/domain/entities/booking.entity.ts
import { Entity } from "@/shared/domain/entities/entity";
import { Actions } from "../enums/actions.enum";
import { Status } from "../enums/status.enum";

export type BookingProps = {
    id: number;
    placedAt: Date;
    start: Date;
    end: Date;
};

export default class BookingEntity extends Entity<BookingProps> {    
    private status: Status

    constructor(public readonly props: BookingProps) {
        super(props);
        this.status = Status.Created;
    }

    currentStatus(): Status {
        return this.status;
    }

    changeState(action: Actions): void {
        switch (this.status) {
            case Status.Created:
                if (action === Actions.Pay) {
                    this.status = Status.Paid;
                } else if (action === Actions.Cancel) {
                    this.status = Status.Canceled;
                }
                break;

            case Status.Paid:
                if (action === Actions.Finish) {
                    this.status = Status.Finished;
                } else if (action === Actions.Refound) {
                    this.status = Status.Refounded;
                }
                break;

            case Status.Canceled:
                if (action === Actions.Reopen) {
                    this.status = Status.Created;
                }
                break;
        }
    }
}